src/pybind/mgr/dashboard/frontend/src/app/shared/directives/tab-stateful.directive.ts

```diff
@@ -0,0 +1,44 @@
+import { Directive, Host, HostListener, Input, OnInit, Optional } from '@angular/core';
+
+import * as _ from 'lodash';
+import { TabDirective } from 'ngx-bootstrap/tabs';
+
+/**
+ * This directive stores the latest selected tab in a tabset and restores its
+ * state on initialization.
+ * Each tab including the parent tabset must have a `stateId` attribute which
+ * is used as an identifier. The current selected tab is stored in local
+ * storage with the key `tabset_<TABSETSTATEID>`and value `<TABSTATEID>`.
+ */
+@Directive({
+  selector: '[cdTabStateful]'
+})
+export class TabStatefulDirective implements OnInit {
+  private localStorage = window.localStorage;
+  private tabsetStateId: string;
+
+  @Input()
+  stateId: string;
+
+  constructor(@Optional() @Host() private tab: TabDirective) {}
+
+  ngOnInit() {
+    const nativeElement = _.get(this.tab.tabset, 'elementRef.nativeElement');
```

Comment:
This way you can read it from the directive attr itself or from heading:
```suggestion
  @Input('cdTabStateful') stateId: string;
  @Input() heading: string;

  constructor(@Optional() @Host() private tab: TabDirective) {}

  ngOnInit() {
    this.stateId = this.stateId || this.heading;
    const nativeElement = _.get(this.tab.tabset, 'elementRef.nativeElement');
```

Reply:
IMO too complicated. I love clean code.

Reply:
`@Input('id')` or `attr.id` doesn't work here? I don't think you need a different input name.

Reply:
Using `id` might be an option instead of a dedicated attribute. I will check that.

Reply:
> If the `id` value is not the empty string, it must be unique in a document (https://developer.mozilla.org/en-US/docs/Web/API/Element/id)

Because of that i think it is better to use a dedicated attribute because 
only the `stateId` for a `tabset` must be unique within a document. The `stateId`for `tab` must only be unique in the context of a `tabset`. We have many `Details` tabs across our UI, thus using `id` will force us to make sure every tab has a unique identifier, which is fairly complicated and prone to errors.

Reply:
Not sure if id/anchors/fragments works the same with Angular as with regular HTML, but if you navigate to a URL#fragment, that fragment needs to be unique too, which on the other hand, makes total sense (how could you point to some element in the DOM if the ID was non unique?).

Reply:
I in favor of implementing the suggestion of @tspmelo or @epuertat. As every tab has a heading it, would be nice to just use that instead of manually adding `heading.toLowerCase().replace(/ /g, '-')`. This will also prevent to forget adding the `stateId` attribute.

If the `tabset` attribute has no `stateId` you could use the URL to determine the path. Something like this is done to determine the selected columns in the table listing (maybe you could use the same calculation and add '_tabs')

Reply:
I stick with my reasoning why I want to keep the previous implementation. I don't want to implement the use of an attribute that has no cross-connection to this feature at all. Many years of programming experience tell me that this only leads to problems.

Reply:
I'm currently testing the usage of `ng-bootstrap` for tabs (and I'm liking this one better than the current one) and with this plugin we no longer use the `heading` attribute.
So I'm OK with how this is implemented.